### ISSUE TITLE
Reduce dependencies

### ISSUE THREAD

**Issue by fregante** [pixiebrix/pixiebrix-extension#411]
Let me know if you'd like me to remove unused dependencies (I see the whole npm is a dependency) and remove unnecessary ones (`deep-equal` has 10 sub-reps because it tries to support ES3 browsers)

**Comment by twschiller** [comment]
That'd be great — should also work to help speed up our CI times (see: #410)

**Comment by fregante** [comment]
Can I also move build/test-related dependencies to `devDependencies`? This way it's a bit easier to see what goes into the bundle (`dependencies`) and what is just used locally.

**Comment by twschiller** [comment]
Absolutely - hadn't prioritized keeping them separate in this repo because there isn't any benefit to build artifact size (or at least there should be). But definitely better to keep separate from a on-boarding, maintenance, and security scanning perspective

**Comment by fregante** [comment]
Why are there 3 slightly different templating engines? Could we drop all but one?

**Comment by twschiller** [comment]
> Why are there 3 slightly different templating engines? Could we drop all but one?

1. Mustache (the default) is nice and safe since it's logic-less
2. Nunjucks is the most expressive
3. Handlebars could potentially get removed at somepoint... I think it's completely subsumed by Nunjucks

For now let's keep all of them. If Handlebars/Nunjucks is large, we can put them behind a dynamic import

**Comment by twschiller** [comment]
Splitting into devDependencies will also be great for cleaning up the `npm audit` output since we can focus on the production dependencies

**Comment by fregante** [comment]
Blocked by #435 and #443, to avoid messy conflicts.

**Comment by fregante** [comment]
Let me know if you'd like me to continue on this issue or if we should pause it for now.

**Comment by twschiller** [comment]
Sure, let's pick off the low hanging issues:

* Let's add lint rules to block the problematic imports. I want to avoid having to manually call out lodash and react-bootrap imports, though. Which imports are the problematic right now? You can see which eslint plugins Codacy support in their package.json: https://github.com/codacy/codacy-eslint. Open a ticket if you need a different one to prevent the problematic imports 
* We can rip out moment.calender and just put date and time (in the user's current time zone). It looks like its only being used in the log display
* I don't think we depend on `pie6k/jquery.initialize` anymore? I had copied it over to our vendors folder to expose it as an es6 module. It's small and IIRC provides functionality beyond MutationObserver (it uses MutationObserver directly if a selector is CSS compatible)

**Comment by fregante** [comment]
> * I don't think we depend on `pie6k/jquery.initialize` anymore?

It seems to still be part of:

https://github.com/pixiebrix/pixiebrix-extension/blob/60189bc8eb4f699f8f7ba979636694c3402d15cc/src/extensionPoints/helpers.ts#L159

I haven't sunk my teeth into the DOM changes detection/injection yet, but I think a lot of code could be replaced by this, which is what GitHub uses: https://github.com/josh/selector-observer

It gives a simple, unified API to track elements that come in and out of a page.

> * I think we could switch over to jquery slim

If jQuery is not used by users, we could definitely trim it down or drop it altogether. It doesn't seem to be used excessively, but dropping it might make you a little less efficient with the DOM if you're not familiar with its raw API.

---

I'd probably look into dropping handlebars sooner rather than later as that seems to be one of the largest dependencies that can probably be replaced by nunchucks (or viceversa)

That large `esprima.js` can also be dropped by updating one of the dependencies, I forgot which one (yaml or one of the templating)

**Comment by twschiller** [comment]
> It seems to still be part of:

I had copied the code into our repository so I could expose it as a module. So if it's still in package.json, we should be able to drop it. The npm version was breaking Jest tests (because the module initialization would fail based on jQuery/the DOM not being available)

> jQuery is not used by users

jQuery-only selectors are used by users including enterprise deployments. E.g., `button:contains("Caption")`. So we won't be able to drop the dependency entirely

> I'd probably look into dropping handlebars sooner rather than later

I'll have to run a check to see if there's any enterprise customer code using it. We should at least make it loaded asynchronously so it doesn't slow down parsing the main bundle

**Comment by fregante** [comment]
> I'm going to go ahead an close this issue for now since the main work was done?

Sure. If there's any specific improvement you's like sooner or later a new issue can be opened (like for Handlebars)

> > It seems to still be part of:
> 
> I had copied the code into our repository so I could expose it as a module

Yes, I was talking about dropping the file in vendors. 



> > jQuery is not used by users
> 
> jQuery-only selectors are used by users

Sizzle, part of jQuery, could potentially be used to only handle the selectors: https://github.com/jquery/sizzle/wiki

**Comment by fregante** [comment]
Just for the record, I may have found a simple solution to see _why_ a certain dependency is being imported. It may not be complete, but it's something:

https://www.npmjs.com/package/webpack-stats-report

```js
// copy-paste into webpack.config.js
const {StatsReportPlugin} = require("webpack-stats-report");

// … 

plugins: [
  new StatsReportPlugin(),
]
```

Example:

<img width="653" alt="Screen Shot 6" src="https://user-images.githubusercontent.com/1402241/154828450-6005e45f-918b-4733-948d-efde249bdae9.png">
